app/src/main/java/com/hedvig/app/feature/insurance/ui/tab/InsuranceViewModel.kt

```diff
@@ -47,17 +50,55 @@ class InsuranceViewModelImpl(
                         data = result.value,
                         showCrossSellNotificationBadge = showNotificationBadge
                     )
-                    _viewState.value = ViewState.Success(items)
+                    _viewState.value = ViewState.Success(items, null)
                 }
             }
         }
     }
 
-    override fun markCardCrossSellsAsSeen() {
+    fun markCardCrossSellsAsSeen() {
         viewModelScope.launch {
             crossSellNotificationBadgeService.markCurrentCrossSellsAsSeen(
                 CrossSellNotificationBadgeService.CrossSellBadgeType.InsuranceFragmentCard
             )
         }
     }
+
+    fun onClickCrossSell(action: CrossSellData.Action) {
+        when (action) {
+            CrossSellData.Action.Chat -> _viewState.update { viewState ->
+                when (viewState) {
+                    ViewState.Error, ViewState.Loading -> viewState
```

Comment:
Does this line do nothing in this case then? Should we think of changing this to something more like:
```kotlin
CrossSellData.Action.Chat -> {
    if (viewState is ViewState.Success) {
        viewState.update { viewState -> viewState.copy(action = action) }
    }
} 
```
to show its intent of doing nothing in those cases?

Reply:
`viewState` can change between these lines though, right? The guard-clause is mostly to make sure that the `viewState` is in a legal state once we want to perform the update. The state will implicitly be correct once the function is called, as it can only be called from a legal state already. Another check would be redundant.

Reply:
Right, fair point. Also while typing it out I see that we don't actually have the guarantee of the type in there since as you also said it could also change after you enter that scope and would have to check again. I totally see what's happening much better now, thanks!